10_rest/kasunfdo/pkg/puppy/store/mapstore.go

```diff
@@ -0,0 +1,78 @@
+package store
+
+import (
+	"sync"
+
+	"github.com/anz-bank/go-course/10_rest/kasunfdo/pkg/puppy"
+)
+
+// MapStore is map based implementation of Storer interface
+type MapStore struct {
+	store map[uint64]puppy.Puppy
+	sync.Mutex
+	nextID uint64
+}
+
+// NewMapStore creates a new MapStore
+func NewMapStore() *MapStore {
+	return &MapStore{nextID: 1, store: map[uint64]puppy.Puppy{}}
+}
+
+// CreatePuppy inserts given puppy in store and returns given id to the puppy
+// Will return an error if value of puppy is negative
+func (m *MapStore) CreatePuppy(p puppy.Puppy) (uint64, error) {
+	if err := p.Validate(); err != nil {
+		return 0, err
+	}
+
+	m.Lock()
+	defer m.Unlock()
+
+	p.ID = m.nextID
+	m.nextID++
+	m.store[p.ID] = p
+	return p.ID, nil
+}
+
+// ReadPuppy reads puppy with given id from the store
+// Will return an error if puppy with given id does not exist
+func (m *MapStore) ReadPuppy(id uint64) (puppy.Puppy, error) {
+	p, ok := m.store[id]
```

Comment:
You're going to need to lock the mutex here too. The map itself is not safe for concurrent access. While this operation is looking up the map, a create may be modifying it. This can cause this lookup to panic or crash because it is not atomic.

This was not needed in the `sync.Map` version because that type is safe for concurrent access.